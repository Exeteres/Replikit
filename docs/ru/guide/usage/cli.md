# Replikit CLI

## Установка

Устанавливать CLI лучше всего глобально:

```shell
npm install -g replikit
# or
yarn global add replikit
```

Тогда вы сможете использовать команду `replikit init` для создания проекта.

## Модули

`Repliki CLI` - это удобный способ создавать модули и управлять ими.
Каждый модуль это полноценный npm проект со своими зависимостями.
Как правило модули не зависят друг от друга.

В `Replikit` выделяется такой вид модулей как `статические`.
Это модули, которые не экспортируют какой-либо функционал для использования в коде,
а просто загружаются и устанавливает хуки, регистрируют контроллеры, обработчики, применяют патчи
или другим образом влияют на работу приложение.
Примерами таких модулей служат контроллеры и `@replikit/attachments`.
Такие модули должны явно подключаться используя ключевое слово `import`.
Лучшее место для их загрузки - файл `replikit.config.ts`,
который будет создан в корне проекта после выполнения `replikit init`.

```ts
import { Configuration } from "@replikit/core";

import "@myproject/mymodule";
import "@replikit/telegram";

const config: Configuration = {};

export default config;
```

Помимо статических модулей `@replikit` здесь также импортируются статические модули из вашего проекта
и экспортируется объект с конфигурацией.

Все остальные модули, такие как `@replikit/router` или `@replikit/storage` должны быть зависимостями отдельных модулей проекта,
а их функционал должен импортироваться и использоваться в коде в соответствии с документацией.

## Доступные команды

| Команда                  | Описание                                 |
| ------------------------ | ---------------------------------------- |
| `replikit init`          | Создает новый репозиторий с модулями     |
| `replikit create-module` | Создает новый модуль в репозитории       |
| `replikit dev`           | Запускает приложение в режиме разработки |
