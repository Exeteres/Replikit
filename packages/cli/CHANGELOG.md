# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.9.0](https://github.com/Exeteres/Replikit/compare/v0.8.0...v0.9.0) (2020-06-28)

**Note:** Version bump only for package @replikit/cli





# [0.8.0](https://github.com/Exeteres/Replikit/compare/v0.7.2...v0.8.0) (2020-06-27)

**Note:** Version bump only for package @replikit/cli





## [0.7.2](https://github.com/Exeteres/Replikit/compare/v0.7.1...v0.7.2) (2020-06-23)

**Note:** Version bump only for package @replikit/cli





# [0.7.0](https://github.com/Exeteres/Replikit/compare/v0.6.0...v0.7.0) (2020-06-23)


### Bug Fixes

* **cli:** Fix generated package.json and tsconfig ([83854b5](https://github.com/Exeteres/Replikit/commit/83854b53c86f691613f42218a8f321270ad22f93))





# [0.6.0](https://github.com/Exeteres/Replikit/compare/v0.5.3...v0.6.0) (2020-06-21)

**Note:** Version bump only for package @replikit/cli





## [0.5.3](https://github.com/Exeteres/Replikit/compare/v0.5.2...v0.5.3) (2020-06-20)

**Note:** Version bump only for package @replikit/cli





## [0.5.2](https://github.com/Exeteres/Replikit/compare/v0.5.1...v0.5.2) (2020-06-20)

**Note:** Version bump only for package @replikit/cli





# 0.5.0 (2020-06-20)


### Bug Fixes

* **cli:** Add @types/node to generated package.json ([4787912](https://github.com/Exeteres/Replikit/commit/4787912620fd2e332728a687ae92f5563fd6f350))
* **cli:** Add missing dependencies ([a5ba2c1](https://github.com/Exeteres/Replikit/commit/a5ba2c1f9ba83692b118655b519d58b14d85f76f))
* **cli:** Fix "workspaces" section in generated package.json ([3dc4214](https://github.com/Exeteres/Replikit/commit/3dc4214a8f929afca7015d9cd3942276ec779c41))
* **cli:** Fix configuration override ([1866a8a](https://github.com/Exeteres/Replikit/commit/1866a8a005055d4a375104964c3d01508ac56cef))


### Features

* **cli:** Add build command ([5c9dc57](https://github.com/Exeteres/Replikit/commit/5c9dc57b9ac2428f4970096dcfc6ae75b2e4dd27))
* **cli:** Add scripts section and replikit as dev dependency to generated package.json ([2f5dd6f](https://github.com/Exeteres/Replikit/commit/2f5dd6f35f47f5fa50109b7155f56109b1ba3efc))






# [0.4.0](https://github.com/Exeteres/Replikit/compare/v0.3.0...v0.4.0) (2020-06-19)


### Features

* **cli:** Add build command ([0212d7f](https://github.com/Exeteres/Replikit/commit/0212d7f9c87d517b8cf220be9b32376551c3603b))





# [0.3.0](https://github.com/Exeteres/Replikit/compare/v0.2.0...v0.3.0) (2020-06-14)


### Bug Fixes

* **cli:** Add @types/node to generated package.json ([3d4fc2b](https://github.com/Exeteres/Replikit/commit/3d4fc2b1c5d0daad6040fe956d3cf164f3b1a227))
* **cli:** Fix configuration override ([d70cddb](https://github.com/Exeteres/Replikit/commit/d70cddb919469e90ed29a1bead411908bf370b57))





# [0.2.0](https://github.com/Exeteres/Replikit/compare/v0.1.0...v0.2.0) (2020-06-13)


### Bug Fixes

* **cli:** Add missing dependencies ([a655059](https://github.com/Exeteres/Replikit/commit/a65505938400bf29c2b890966a2685ac7618d989))
* **cli:** Fix "workspaces" section in generated package.json ([12aca8b](https://github.com/Exeteres/Replikit/commit/12aca8b5f729834c270abfed5606e932b987d48b))


### Features

* **cli:** Add scripts section and replikit as dev dependency to generated package.json ([64399a2](https://github.com/Exeteres/Replikit/commit/64399a242938b1e12baef40a75cf0d6cfeec3b0f))
