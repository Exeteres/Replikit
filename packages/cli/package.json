{
    "name": "@replikit/cli",
    "version": "0.5.3",
    "license": "MIT",
    "main": "dist/index.js",
    "typings": "dist/index.d.ts",
    "files": [
        "dist",
        "typings"
    ],
    "scripts": {
        "build": "ttsc -p tsconfig.build.json",
        "dev": "ts-node -r tsconfig-paths/register src/main.ts"
    },
    "bin": {
        "replikit": "./dist/main.js"
    },
    "dependencies": {
        "@replikit/chalk": "^0.5.3",
        "@replikit/core": "^0.5.3",
        "commander": "^5.1.0",
        "dotenv": "^8.2.0",
        "execa": "^4.0.0",
        "fs-extra": "^9.0.0",
        "inquirer": "^7.1.0",
        "native-addon-loader": "^2.0.1",
        "require-from-string": "^2.0.2",
        "ts-loader": "^7.0.5",
        "ts-morph": "^7.0.3",
        "ts-node": "^8.0.2",
        "ts-node-dev": "^1.0.0-pre.44",
        "tsconfig-paths": "^3.9.0",
        "ttypescript": "^1.5.10",
        "typescript": "^3.8.3",
        "webpack": "^4.43.0",
        "webpack-virtual-modules": "^0.2.2",
        "webpackbar": "^4.0.0"
    },
    "devDependencies": {
        "@types/fs-extra": "^8.1.0",
        "@types/inquirer": "^6.5.0",
        "@types/require-from-string": "^1.2.0",
        "@types/webpack": "^4.41.17",
        "@types/webpack-virtual-modules": "^0.1.0",
        "@types/webpackbar": "^4.0.0"
    }
}
