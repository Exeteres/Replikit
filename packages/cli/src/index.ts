/// <reference path="../typings/index.d.ts" />

export * from "./controllers/shellController";
export * from "./controllers/gitController";
export * from "./controllers/pm/pmType";
export * from "./controllers/pm/pmController";
export * from "./controllers/pm/npmController";
export * from "./controllers/pm/yarnController";
export * from "./controllers/pm/pmDetector";

export * from "./project/moduleManager";
export * from "./project/projectManager";
export * from "./project/configManager";

export * from "./actions/initProject";
export * from "./actions/createModule";

export * from "./errors";
export * from "./utils";
export * from "./startup";
export * from "./webpack";

export * from "./shared";

import "./args";

import "./commands/init";
import "./commands/createModule";
import "./commands/dev";
import "./commands/build";
import "./commands/addExternal";
