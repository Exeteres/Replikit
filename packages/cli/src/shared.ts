/** @internal */
export const availableModules = [
    "@replikit/router",
    "@replikit/messages",
    "@replikit/commands",
    "@replikit/storage",
    "@replikit/i18n",
    "@replikit/permissions"
];

/** @internal */
export const availableStaticModules = [
    "@replikit/telegram",
    "@replikit/vk",
    // "@replikit/discord",
    "@replikit/attachments",
    "@replikit/authorization"
];
