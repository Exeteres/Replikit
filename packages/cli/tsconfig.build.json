{
    "extends": "../../tsconfig.build.json",
    "compilerOptions": {
        "outDir": "dist",
        "paths": {
            "@replikit/cli": ["cli/src"]
        }
    },
    "include": ["src"]
}
