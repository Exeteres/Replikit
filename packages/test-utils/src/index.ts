export * from "./databaseTestManager";
export * from "./tmp";
export * from "./code";
export * from "./expect";
export * from "./testController";
export * from "./commandTestManager";
export * from "./factories";
