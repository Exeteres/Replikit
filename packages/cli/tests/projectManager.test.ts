import { setupTemp, expectDirectoryToMatchSnapshot } from "@replikit/test-utils";
import { createModule, ProjectManager, GitController, PMType } from "@replikit/cli";
import { createProject } from "./shared";
import { resolve } from "path";

setupTemp();

jest.setTimeout(120000);

describe("ProjectManager", () => {
    it.each(["npm", "yarn", "yarn-lerna"])("should generate %s project", async type => {
        const useYarn = type !== "npm";
        const useLerna = type !== "yarn";

        const manager = await createProject(useYarn, useLerna);
        await expectDirectoryToMatchSnapshot(manager.root);
    });

    it.each(["with", "without"])("should add a module to the project %s logger", async t => {
        const addLoger = t === "with";
        const manager = await createProject(false, false);
        await createModule(manager, "test-module", ["@replikit/test"], addLoger);
        await expectDirectoryToMatchSnapshot(manager.root);
    });

    it("should add external subrepository, update tsconfig.json and add modules to replikit config", async () => {
        const gitSpy = jest.spyOn(GitController.prototype, "addSubmodule");
        gitSpy.mockImplementation(async function (this: GitController, _, folder) {
            const externalManager = new ProjectManager(resolve(this.path, folder));
            externalManager.setPackageManager(PMType.NPM);
            await externalManager.init();
            await externalManager.createModule("external-module");
        });

        const manager = await createProject(false, false);
        await manager.git.addSubmodule(undefined!, "external-repo");
        await manager.addExternalRepo("external-repo");

        const modules = await manager.getExternalModuleNames("external-repo");
        await manager.addLocalModules(modules);
        await expectDirectoryToMatchSnapshot(manager.root);

        gitSpy.mockRestore();
    });
});
