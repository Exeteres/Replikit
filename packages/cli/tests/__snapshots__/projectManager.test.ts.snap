// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectManager should add a module to the project with logger: /modules/test-module/package.json 1`] = `
"{
    \\"name\\": \\"@test-project/test-module\\",
    \\"version\\": \\"0.0.0\\",
    \\"license\\": \\"MIT\\",
    \\"dependencies\\": {
        \\"@replikit/test\\": \\"0.0.0\\"
    }
}
"
`;

exports[`ProjectManager should add a module to the project with logger: /modules/test-module/src/index.ts 1`] = `
"export * from \\"./startup\\";
"
`;

exports[`ProjectManager should add a module to the project with logger: /modules/test-module/src/startup.ts 1`] = `
"import { createScope } from \\"@replikit/core\\";

/** @internal */
export const logger = createScope(\\"test-module\\");
"
`;

exports[`ProjectManager should add a module to the project with logger: /package.json 1`] = `
"{
    \\"name\\": \\"test-project\\",
    \\"version\\": \\"0.0.0\\",
    \\"license\\": \\"MIT\\",
    \\"private\\": true,
    \\"scripts\\": {
        \\"dev\\": \\"replikit dev\\",
        \\"build\\": \\"replikit build\\"
    },
    \\"dependencies\\": {
        \\"@replikit/static\\": \\"0.0.0\\"
    },
    \\"devDependencies\\": {
        \\"@types/node\\": \\"0.0.0\\",
        \\"replikit\\": \\"0.0.0\\"
    }
}
"
`;

exports[`ProjectManager should add a module to the project with logger: /replikit.config.ts 1`] = `
"import { Configuration } from \\"@replikit/core/typings\\";

import \\"@replikit/static\\";
import \\"@test-project/test-module\\";

const config: Configuration = {};

export default config;
"
`;

exports[`ProjectManager should add a module to the project with logger: /tsconfig.json 1`] = `
"{
    \\"compilerOptions\\": {
        \\"target\\": \\"es2019\\",
        \\"module\\": \\"commonjs\\",
        \\"moduleResolution\\": \\"node\\",
        \\"baseUrl\\": \\"modules\\",
        \\"strict\\": true,
        \\"paths\\": {
            \\"@test-project/*\\": [
                \\"*/src\\"
            ],
            \\"@test-project/*/typings\\": [
                \\"*/typings\\"
            ]
        }
    },
    \\"include\\": [
        \\"replikit.config.ts\\",
        \\"modules/**/*.ts\\"
    ]
}
"
`;

exports[`ProjectManager should add a module to the project without logger: /modules/test-module/package.json 1`] = `
"{
    \\"name\\": \\"@test-project/test-module\\",
    \\"version\\": \\"0.0.0\\",
    \\"license\\": \\"MIT\\",
    \\"dependencies\\": {
        \\"@replikit/test\\": \\"0.0.0\\"
    }
}
"
`;

exports[`ProjectManager should add a module to the project without logger: /modules/test-module/src/index.ts 1`] = `""`;

exports[`ProjectManager should add a module to the project without logger: /package.json 1`] = `
"{
    \\"name\\": \\"test-project\\",
    \\"version\\": \\"0.0.0\\",
    \\"license\\": \\"MIT\\",
    \\"private\\": true,
    \\"scripts\\": {
        \\"dev\\": \\"replikit dev\\",
        \\"build\\": \\"replikit build\\"
    },
    \\"dependencies\\": {
        \\"@replikit/static\\": \\"0.0.0\\"
    },
    \\"devDependencies\\": {
        \\"@types/node\\": \\"0.0.0\\",
        \\"replikit\\": \\"0.0.0\\"
    }
}
"
`;

exports[`ProjectManager should add a module to the project without logger: /replikit.config.ts 1`] = `
"import { Configuration } from \\"@replikit/core/typings\\";

import \\"@replikit/static\\";
import \\"@test-project/test-module\\";

const config: Configuration = {};

export default config;
"
`;

exports[`ProjectManager should add a module to the project without logger: /tsconfig.json 1`] = `
"{
    \\"compilerOptions\\": {
        \\"target\\": \\"es2019\\",
        \\"module\\": \\"commonjs\\",
        \\"moduleResolution\\": \\"node\\",
        \\"baseUrl\\": \\"modules\\",
        \\"strict\\": true,
        \\"paths\\": {
            \\"@test-project/*\\": [
                \\"*/src\\"
            ],
            \\"@test-project/*/typings\\": [
                \\"*/typings\\"
            ]
        }
    },
    \\"include\\": [
        \\"replikit.config.ts\\",
        \\"modules/**/*.ts\\"
    ]
}
"
`;

exports[`ProjectManager should generate npm project: /lerna.json 1`] = `
"{
    \\"version\\": \\"0.0.0\\",
    \\"npmClient\\": \\"npm\\"
}
"
`;

exports[`ProjectManager should generate npm project: /package.json 1`] = `
"{
    \\"name\\": \\"test-project\\",
    \\"version\\": \\"0.0.0\\",
    \\"license\\": \\"MIT\\",
    \\"private\\": true,
    \\"scripts\\": {
        \\"dev\\": \\"replikit dev\\",
        \\"build\\": \\"replikit build\\"
    },
    \\"dependencies\\": {
        \\"@replikit/static\\": \\"0.0.0\\"
    },
    \\"devDependencies\\": {
        \\"@types/node\\": \\"0.0.0\\",
        \\"replikit\\": \\"0.0.0\\",
        \\"lerna\\": \\"0.0.0\\"
    }
}
"
`;

exports[`ProjectManager should generate npm project: /replikit.config.ts 1`] = `
"import { Configuration } from \\"@replikit/core/typings\\";

import \\"@replikit/static\\";

const config: Configuration = {};

export default config;
"
`;

exports[`ProjectManager should generate npm project: /tsconfig.json 1`] = `
"{
    \\"compilerOptions\\": {
        \\"target\\": \\"es2019\\",
        \\"module\\": \\"commonjs\\",
        \\"moduleResolution\\": \\"node\\",
        \\"baseUrl\\": \\"modules\\",
        \\"strict\\": true,
        \\"paths\\": {
            \\"@test-project/*\\": [
                \\"*/src\\"
            ],
            \\"@test-project/*/typings\\": [
                \\"*/typings\\"
            ]
        }
    },
    \\"include\\": [
        \\"replikit.config.ts\\",
        \\"modules/**/*.ts\\"
    ]
}
"
`;

exports[`ProjectManager should generate yarn project: /package.json 1`] = `
"{
    \\"name\\": \\"test-project\\",
    \\"version\\": \\"0.0.0\\",
    \\"license\\": \\"MIT\\",
    \\"private\\": true,
    \\"scripts\\": {
        \\"dev\\": \\"replikit dev\\",
        \\"build\\": \\"replikit build\\"
    },
    \\"workspaces\\": [
        \\"modules/*\\"
    ],
    \\"dependencies\\": {
        \\"@replikit/static\\": \\"0.0.0\\"
    },
    \\"devDependencies\\": {
        \\"@types/node\\": \\"0.0.0\\",
        \\"replikit\\": \\"0.0.0\\"
    }
}
"
`;

exports[`ProjectManager should generate yarn project: /replikit.config.ts 1`] = `
"import { Configuration } from \\"@replikit/core/typings\\";

import \\"@replikit/static\\";

const config: Configuration = {};

export default config;
"
`;

exports[`ProjectManager should generate yarn project: /tsconfig.json 1`] = `
"{
    \\"compilerOptions\\": {
        \\"target\\": \\"es2019\\",
        \\"module\\": \\"commonjs\\",
        \\"moduleResolution\\": \\"node\\",
        \\"baseUrl\\": \\"modules\\",
        \\"strict\\": true,
        \\"paths\\": {
            \\"@test-project/*\\": [
                \\"*/src\\"
            ],
            \\"@test-project/*/typings\\": [
                \\"*/typings\\"
            ]
        }
    },
    \\"include\\": [
        \\"replikit.config.ts\\",
        \\"modules/**/*.ts\\"
    ]
}
"
`;

exports[`ProjectManager should generate yarn-lerna project: /lerna.json 1`] = `
"{
    \\"version\\": \\"0.0.0\\",
    \\"npmClient\\": \\"yarn\\",
    \\"useWorkspaces\\": true
}
"
`;

exports[`ProjectManager should generate yarn-lerna project: /package.json 1`] = `
"{
    \\"name\\": \\"test-project\\",
    \\"version\\": \\"0.0.0\\",
    \\"license\\": \\"MIT\\",
    \\"private\\": true,
    \\"scripts\\": {
        \\"dev\\": \\"replikit dev\\",
        \\"build\\": \\"replikit build\\"
    },
    \\"workspaces\\": [
        \\"modules/*\\"
    ],
    \\"dependencies\\": {
        \\"@replikit/static\\": \\"0.0.0\\"
    },
    \\"devDependencies\\": {
        \\"@types/node\\": \\"0.0.0\\",
        \\"replikit\\": \\"0.0.0\\",
        \\"lerna\\": \\"0.0.0\\"
    }
}
"
`;

exports[`ProjectManager should generate yarn-lerna project: /replikit.config.ts 1`] = `
"import { Configuration } from \\"@replikit/core/typings\\";

import \\"@replikit/static\\";

const config: Configuration = {};

export default config;
"
`;

exports[`ProjectManager should generate yarn-lerna project: /tsconfig.json 1`] = `
"{
    \\"compilerOptions\\": {
        \\"target\\": \\"es2019\\",
        \\"module\\": \\"commonjs\\",
        \\"moduleResolution\\": \\"node\\",
        \\"baseUrl\\": \\"modules\\",
        \\"strict\\": true,
        \\"paths\\": {
            \\"@test-project/*\\": [
                \\"*/src\\"
            ],
            \\"@test-project/*/typings\\": [
                \\"*/typings\\"
            ]
        }
    },
    \\"include\\": [
        \\"replikit.config.ts\\",
        \\"modules/**/*.ts\\"
    ]
}
"
`;
